packages/eslint-plugin/src/rules/require-types-exports.ts

```diff
@@ -0,0 +1,403 @@
+            member.typeAnnotation?.typeAnnotation.type ===
+              AST_NODE_TYPES.TSTypeReference
+          ) {
+            acc.push(member.typeAnnotation.typeAnnotation);
+          }
+
+          return acc;
+        }, []);
+      }
+
+      return [];
+    }
+
+    function convertGenericTypeToTypeReference(
+      functionNode: FunctionNode,
+      typeNode: TSESTree.TSTypeReference,
+    ): TSESTree.TSTypeReference | TSESTree.TSTypeReference[] {
+      const typeName = getTypeName(typeNode);
+
+      if (!typeName) {
+        return typeNode;
+      }
+
+      const scope = context.sourceCode.getScope(functionNode);
+      const variable = scope.set.get(typeName);
+
+      if (!variable?.isTypeVariable) {
+        return typeNode;
+      }
+
+      for (const definition of variable.defs) {
+        if (
+          definition.type === DefinitionType.Type &&
+          definition.node.type === AST_NODE_TYPES.TSTypeParameter &&
+          definition.node.constraint
+        ) {
+          switch (definition.node.constraint.type) {
+            // T extends SomeType
+            case AST_NODE_TYPES.TSTypeReference:
+              return definition.node.constraint;
+
+            // T extends SomeType | AnotherType
+            // T extends SomeType & AnotherType
+            case AST_NODE_TYPES.TSUnionType:
+            case AST_NODE_TYPES.TSIntersectionType:
+              return definition.node.constraint.types.filter(
+                type => type.type === AST_NODE_TYPES.TSTypeReference,
+              ) as TSESTree.TSTypeReference[];
+
+            // T extends [SomeType, AnotherType]
+            case AST_NODE_TYPES.TSTupleType:
+              return definition.node.constraint.elementTypes.filter(
+                type => type.type === AST_NODE_TYPES.TSTypeReference,
+              ) as TSESTree.TSTypeReference[];
+
+            // T extends { some: SomeType, another: AnotherType }
+            case AST_NODE_TYPES.TSTypeLiteral:
+              return definition.node.constraint.members.reduce<
+                TSESTree.TSTypeReference[]
+              >((acc, member) => {
+                if (
+                  member.type === AST_NODE_TYPES.TSPropertySignature &&
+                  member.typeAnnotation?.typeAnnotation.type ===
+                    AST_NODE_TYPES.TSTypeReference
+                ) {
+                  acc.push(member.typeAnnotation.typeAnnotation);
+                }
+
+                return acc;
+              }, []);
+
+            default:
+              continue;
+          }
+        }
+      }
+
+      return typeNode;
+    }
+
+    function getTypeName(typeReference: TSESTree.TSTypeReference): string {
+      if (typeReference.typeName.type === AST_NODE_TYPES.Identifier) {
+        return typeReference.typeName.name;
+      }
+
+      return '';
+    }
+
+    return {
+      'ImportDeclaration[importKind="type"] ImportSpecifier, ImportSpecifier[importKind="type"]':
+        collectImportedTypes,
```

Comment:
Currently, it doesn't support this:

```ts
import { Type } from './types';

export function (arg: Type) { /* ... */ }
```

but only things like this:

```ts
import type { Type1 } from './types';
import { type Type2 } from './types';
import { type Type3 as Type4 } from './types';
```

because IDK if I can know for sure if the imported name is a type or not
any idea if this could be done?

Reply:
The type checker can determine whether something is in type space and/or value space.

But for that case of `export function (arg: Type) { /* ... */ }`, do we need to know whether it's type and/or value? I'd say that should be a complaint in the rule no matter what, no?

Reply:
Done

Reply:
OK,
After thinking about it for a little bit...

What should we do in cases where someone imports a type from another module and uses it in their exported function?

```TS
// my-package/src/types.ts
export type A = number;
```

```TS
// my-package/src/function.ts
import type { A } from './types';

export function f(): A {
	return 1;
}
```

```TS
// my-package/src/index.ts
export { f } from './function';
```

Then, in user-land:
```TS
// my-code.ts
import { f } from 'my-package';

let value; // What type is it?
value = f();
```

I mean... on the one hand, we can't know whether the type is exported somehow to the user.
On the other hand, (I think?) we don't want to enforce everyone to (re)export the types they imported 😓

Reply:
On a [pure scope analysis level](https://typescript-eslint.io/packages/scope-manager), we do know this. The `A` identifier in `function.ts` is part of what's exported with `f()`.

I think it'd be very preferable to stick with the pure scope analysis strategy. Dipping into types land is much harder to work with, since `type` aliases sometimes become new things and sometimes are just references to existing things.

Does that answer what you're looking for?

Reply:
I think the link here is wrong?

Reply:
I meant to link to https://typescript-eslint.io/packages/scope-manager/. As in, the thing you referenced using in another comment.

Reply:
Oh,
so I don't understand what you meant in the previous comment 🫣

Reply:
Oh, sorry! I'm being unclear. Answering more specifically:

> I mean... on the one hand, we can't know whether the type is exported somehow to the user.

I believe we can. In the case of a function like `export function f(): A`, we know that it's exported and that within its return type exists an identifier `A`. The rule would need to check: for every identifier in the return type, is that also exported in the current file (module)?

> On the other hand, (I think?) we don't want to force everyone to (re)export the types they imported

I think we do, and that that's the goal of this rule! In the case of the `f(): A`, it'd be something like:

```ts
// my-package/src/function.ts
import type { A } from './types';

+ export { A };

export function f(): A {
	return 1;
}
```

That does leave a "hole" in the rule, that `my-package/src/index.ts` exports `f` but not `A`. In theory we could dip into the type checker to know this, and then report... but really that's (a) a lot of work and (b) dipping into TypeScript's ["inferred type cannot be named without a reference"](https://stackoverflow.com/questions/72041763/typescript-inferred-type-cannot-be-named-without-reference) territory. I'll defer to @typescript-eslint/triage-team on this one, but I'm ok with the rule not using type checking.

Something like [Knip](https://knip.dev) would IMO be better suited to cross-file type checking things.

Does that answer what you're looking for?

Reply:
> The rule would need to check: for every identifier in the return type, is that also exported in the current file (module)?

My current logic is:
"Does this type exist in the outside world (whether it's imported/exported in the current file)?"
	- If so: Great!
	- Otherwise: Report

> I think we do, and that that's the goal of this rule! 

I don't fully agree
I mean... yeah, that's the goal of the rule, but the type might already be exported through the index file, for example, and we'll force developers to export it multiple times in their code.

Let's take my example above and modify it a little bit:

```TS
// my-package/src/types.ts
export type A = number;
```

```TS
// my-package/src/f1.ts
import type { A } from './types';

export function f1(): A {
	return 1;
}
```

```TS
// my-package/src/f2.ts
import type { A } from './types';

export function f2(): A {
	return 2;
}
```

```TS
// my-package/src/index.ts
export { f1 } from './f1';
export { f2 } from './f2';

export * from './types';
```

Then, in user-land:
```TS
// my-code.ts
import { f1 } from 'my-package';
import type { A } from 'my-package';

let value: A; // I know what the type is! YAY!
value = f1();
```

This code is totally fine. But with the changes you suggest, it'll have 2 more redundant exports (and it'll increase linearly based on the type usage inside other modules).
Yeah, it probably won't hurt anybody, but it will make the DX awful, don't you think?

Reply:
If we don't enforce that a `export function fn(): A;` is in a file that also exports `A`, what is the point of this rule?

Reply:
For lint rules, the line generally lies with the single file being linted. Because lint rules operate on a single file at a time, that naturally falls out of the architecture.

It's pretty common that lint rules will only be able to catch issues within the small scope of a single rule at a time. For example, `no-unused-vars` won't detect things that are exported from a file but never imported. Hence [Knip](https://knip.dev).

Reply:
do you wanna loop in the triage team so we can agree on something together?

Reply:
Already did 🙂 just waiting for them to respond.